Loki: Remove unnecessary deduplication

ivanahuckova opened the pull request:
**What this PR does / why we need it**:
Removes unnecessary deduplication. Read more here https://github.com/grafana/grafana/issues/26714. 
**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana/issues/26714

**Special notes for your reviewer**:

ivanahuckova committed:
Remove unnecessary deduplication

ivanahuckova committed:
Merge remote-tracking branch 'origin' into ivana/remove-deduplication

ivanahuckova committed:
Remove dedup test as we are not doing dedup on all logs  anymore

ivanahuckova commented:
This is a tricky one. Please read for more context. 

Based on communication in https://github.com/grafana/grafana/issues/26714 we've learnt, that we are currently deduplicating Loki logs in Grafana that are not supposed to be duplicated. Therefore this PR removed the deduplication. By removing the deduplication, we have a new problem. We are currently creating/setting each log's `uid` which is basically hashed logs nano-timestamp, labels and log message. However, if we don't deduplicate logs, we can have multiple logs with the same uid and uid is not unique anymore. This **is a problem only in live tailing** where we have animation for new logs that relies on unique keys. But because at first, we are showing logs that were received from the regular range query we need to keep in mind that for the each log has to always have the same uid (no matter in which query it arrives).  This means that this uid can't be created randomly trough uid, or by using index. 

See here how the same log lines have different index, until we start to show only logs received from the websocket:
![image](https://user-images.githubusercontent.com/30407135/101508367-dc9b7880-3977-11eb-9c1f-28bafee619e7.png)


**Couple of solutions that I could think of:**
1. Currently, when live-tailing, we firstly show logs that were received from the regular range query. If we removed this and only shown logs received from the websocket, we could simply remove uid and rely just on index. This would solve the problem and wouldn't have too negative impact on users as it takes less than 1 sec for logs to appear. We could add spinner or something else to show users that logs are being fetched. See below example. If we added spinner or message, I think it wouldn't be too bad regression. 

![Kapture 2020-12-08 at 17 13 55](https://user-images.githubusercontent.com/30407135/101509255-df4a9d80-3978-11eb-822d-b3d8b8aa3395.gif)


2. We would keep the uid and wouldn't do any changes. In an edge case when users receive identical logs (that have the same nano timestamp + log lines + labels), the new logs highlighting would stop to work, but live-tailing and everything would work as before.  Only the animation would be affected. 

@aocenas @davkal thoughts on this?

ivanahuckova commented:
Based on coversation with @davkal and @aocenas, for now, we are going to go with option 2. This PR is then ready for review.

ivanahuckova commented on packages/grafana-ui/src/components/Logs/LogRows.tsx:
```diff
@@ -137,7 +137,7 @@ class UnThemedLogRows extends PureComponent<Props, State> {
               firstRows.map((row, index) => (
                 <LogRow
-                  key={row.uid}
+                  key={`${row.uid}-${index}`}
```
Uid is created by hashing all provided logs information. However, we might have multiple identical logs.Therefore we need to add index here. This pattern is used at other places in code base as well.

aocenas replied:
hmm what I don't like with this is that we have a property uid, which usually means unique identifier and that right away say, this may not be unique. From the meaning of that property, the data source should provide it and if it doesn't we do some fall back to something when processing the logs (I think we already just add an index there or something). So this does not feel right I think we should either abandon creation of uid in the Loki DS because it does not actually provide uids or maybe do an uuid4 or something like that in the datasource.

ivanahuckova commented on public/app/features/explore/Explore.tsx:
```diff
@@ -57,8 +57,8 @@ import { FILTER_FOR_OPERATOR, FILTER_OUT_OPERATOR, FilterItem } from '@grafana/u
   return {
-    logsMain: css`
-      label: logsMain;
+    exploreMain: css`
```
Not related to deduplication, but a little extra fix. 🙂

ivanahuckova commented on public/app/features/explore/LiveLogs.tsx:
```diff
@@ -14,7 +14,7 @@ const getStyles = (theme: GrafanaTheme) => ({
     display: flex;
     flex-flow: column nowrap;
-    height: 65vh;
+    height: 60vh;
```
We have added second line in Loki query editor. This improves UI a little bit.

ivanahuckova committed:
Create unique ids in Loki

ivanahuckova committed:
Fix comment

ivanahuckova committed:
Fix comment

ivanahuckova commented:
After conversation and great tip from @aocenas, I have implemented different solution. In Loki, I have added a logic that checks if the uid was already assigned, and if so, it adds the index after it. This way, uids are truly unique and different queries (range/live tailing) is going to provide consistent results. I have added tests as well.

aocenas commented on public/app/plugins/datasource/loki/result_transformer.ts:
```diff
@@ -127,6 +130,9 @@ export function appendResponseToBufferedData(response: LokiTailResponse, data: M
   const idField = data.fields[4];
 
+  // We need to store and track all used uids to ensure that uids are unique
+  const usedUids: { string?: number } = {};
```
This won't probably work the same way as in `lokiStreamResultToDataFrame` as this is called only on the new lines. We probably have take all the current ids in `data` an use that as a map, otherwise we would be checking here only the few new lines that came later. I mean realistically this could be a problem only if the same line came in 2 separate batches from the websocket not sure how likely that is. What will happen in case we have 2 same ids when livestreaming? If it's only the visual fade highlight maybe I would not just add a comment here that this could happen but would not complicate the code right now.

aocenas reviewed (approved):
Apart from that one comment this looks good

ivanahuckova committed:
Store prev response uids in usedUids

ivanahuckova committed:
Revert "Store prev response uids in usedUids"

This reverts commit 05c496e2a8150573513f2574cfef1407de96a72b.

ivanahuckova committed:
Add comment
